### ISSUE TITLE
Release Twisted 15.2.1

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7906]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7906|
|Type|task|
|Created|2015-05-20 04:59:33Z|
|Branch|https://github.com/twisted/twisted/tree/release-15.2.1-noconflict-7906-2|

A bugfix release for 15.2

<details><summary>Searchable metadata</summary>

```
trac-id__7906 7906
type__task task
reporter__hawkowl hawkowl
priority__normal normal
milestone__ 
branch__branches_release_15_2_1_noconflict_7906_2 branches/release-15.2.1-noconflict-7906-2
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1432097973790142 1432097973790142
changetime__1433857118886845 1433857118886845
version__None None
owner__hawkowl hawkowl
cc__radix
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @hawkowl|
|-|-|

I assume that this is the diff - https://github.com/twisted/twisted/compare/releases/release-15.2.0-7844...releases/release-15.2.1-7906

not sure what is the process of merging these changes in trunk... I could not find info about patch releases at https://twistedmatrix.com/trac/wiki/ReleaseProcess 

Hawkie, can you please provide some guidance for how we should review this patch?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl removed owner|
|-|-|

I think as far as what needs to be done for this branch:

- A reviewer verifies the newsfiles are sane
- A reviewer verifies the builds are acceptable (except debian8-py27, that is only supported in trunk)
- A reviewer verifies that the two tickets in the newsfile ([#7903](https://github.com/twisted/twisted/issues/7903) [#7904](https://github.com/twisted/twisted/issues/7904)) were correctly cherry picked into the release branch.

https://github.com/twisted/twisted/compare/releases/release-15.2.0-7844...releases/release-15.2.1-7906 does seem to be the correct diff :) Once this is verified as something that works, it'll be documented in the release process doc.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

I wanted to understand how patch release are handled.

Here is my current understanding.

Twisted only support a single release.
Once 13.2.0 is release there will be no patch release for a 13.1.X versions or any earlier version.

All bugfixes will first go to trunk.

For important bugs present in the latest release a patch release is done.

The patch release is similar to a hot fix / cumulative fixes and is done in order to speed up the delivery of those bug fixed
... as otherwise the users will need to wait for a new major release in order to get those fixes.

It will be forked from latest release (and not latest trunk) and will cherry pick the bugs.

The release script is executed on the new patch release branch and release archives are generated from it.

Once the patch release is released the news files are merged in trunk. Changes are already in trunk.

Do we need a new ticket to merge the patch release newsfile in trunk ? 

-------

Hawkie, if you have time please update the wiki page to discuss this process.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

Hi Adi,

I think that explains it well!

This is the ticket that will merge the patch release newsfile into trunk -- the rest of it will be no-ops as the changes are already there (so the merge into trunk will be newsfiles + version files, with no other changes).

I'll update the wiki :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @hawkowl|
|-|-|

I see that the current branch can not be merged into trunk without conflicts.

The git diff does not help, but I tried a local squash merged into trunk to check the result of landing this branch into trunk.

It failed only due to newsfiles conflicts.

Still, after the merge the twisted/topfiles contains the 7904 and 7903 files.

I think that the topfiles should also be removed so that they will no be present in a future release.

The topfiles are not present in the patch release branch (https://github.com/twisted/twisted/tree/releases/release-15.2.1-7906/twisted/topfiles) ... most probably since the release-15.2.1 was branched from release-15.2.0 and not from trunk.

I don't know how can we instruct the VCS to delete them on merge.


------


Regarding the changes from https://twistedmatrix.com/trac/wiki/ReleaseProcess?action=diff&version=196&old_version=194

your changes refer the release as 'bug fix' but in the same document it is referred as patch relese... while in twisted.python.versions.Version (https://twistedmatrix.com/documents/current/api/twisted.python.versions.Version.html) it is refered as major.minor.micro .

The wiki changes does not inform that the 'affected version' is always the latest release and we will never have patch released for older releases.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

To merge the patch release into master I think that you can create a new branch from trunk, merge changes from releases/release-15.2.1-7906 into it and manually remove the topfiles.

Once the release and merge is over we can update the wiki page to explain the whole patch release process.

Many thanks for your help with the release!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl commented|
|-|-|

(In [45002]) Branching to release-15.2.1-noconflict-7906-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl removed owner|
|-|-|

I have done a noconflict branch that should merge into trunk with the desired effects.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

Tests green so far.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Latest branch look good. Thanks! Please merge.

Can you please review the current bugfix release process and make sure it includes all the required information about the bugfix release?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45015]) Merge release-15.2.1-noconflict-7906-2: Release Twisted 15.2.1

Author: hawkowl
Reviewer: adiroiban
Fixes: [#7906](https://github.com/twisted/twisted/issues/7906)
